src/Algebra/Bundles/Free/Magma.agda

```diff
@@ -0,0 +1,443 @@
+  isMagma = record { isEquivalence = preservesEquivalence isEqᴬ ; ∙-cong = _∙_ }
+
+  freeMagma : Magma a (a ⊔ ℓa)
+  freeMagma = record { Carrier = Syntax Carrier; _≈_ = _≈ᵀ_ ; _∙_ = _∙_ ; isMagma = isMagma }
+
+-- re-export some substructure
+
+  open Magma freeMagma public using (rawMagma; setoid; Carrier; _≈_)
+
+  varSetoidHomomorphism : SetoidHomomorphism 𝓐 setoid
+  varSetoidHomomorphism = record { ⟦_⟧ = var; isRelHomomorphism = varIsRelHomomorphism }
+
+
+------------------------------------------------------------------------
+-- Semantics: in terms of concrete Magma instances
+
+module _ (𝓜 : Magma m ℓm) where
+
+  open Magma 𝓜 renaming (setoid to setoidᴹ; Carrier to UM; _∙_ to _∙ᴹ_)
+  open Syntax
+
+------------------------------------------------------------------------
+-- Eval, as the unique fold ⟦_⟧_ over Syntax
+
+  module Eval (𝓐 : Setoid a ℓa) where
+
+    open Setoid 𝓐 renaming (Carrier to UA)
+
+    ⟦_⟧_ : Syntax UA → (UA → UM) → UM
+    ⟦ var a ⟧ η = η a
+    ⟦ s ∙ t ⟧ η = ⟦ s ⟧ η ∙ᴹ ⟦ t ⟧ η
+
+------------------------------------------------------------------------
+-- Any Magma *is* an algebra for the Syntax Functor
+  
+  alg : Syntax UM → UM
+  alg t = ⟦ t ⟧ id where open Eval setoidᴹ
+
+------------------------------------------------------------------------
+-- ⟦_⟧_ defines the (unique) lifting of Setoid homomorphisms to Magma homomorphisms
+
+module LeftAdjoint {𝓐 : Setoid a ℓa} (𝓜 : Magma m ℓm)
+       (𝓗 :  SetoidHomomorphism 𝓐 (Magma.setoid 𝓜)) where
+
+  open Magma 𝓜
+    renaming (Carrier to UM; _≈_ to _≈ᴹ_; _∙_ to _∙ᴹ_
+             ; setoid to setoidᴹ; rawMagma to rawMagmaᴹ
+             ; isMagma to isMagmaᴹ)
+
+  open ≈-Reasoning setoidᴹ
+
+  open Syntax
+
+  open Setoid 𝓐 renaming (Carrier to UA; _≈_ to _≈ᴬ_)
+
+  open Eval 𝓜 𝓐 public
+
+  open FreeMagma 𝓐 renaming (setoid to FA; Carrier to UFA)
+
+  open SetoidHomomorphism 𝓗 renaming (⟦_⟧ to η; isRelHomomorphism to hom-η) 
+
+  private
+  
+    ⟦_⟧ᴹ : UFA → UM
+    ⟦_⟧ᴹ = ⟦_⟧ η
+
+  open Structures _≈ᴹ_
+  open IsMagma isMagmaᴹ renaming (∙-cong to congᴹ)
+  open IsRelHomomorphism hom-η renaming (cong to cong-η)
+    
+  module Existence where
+  
+    private
+      algᴹ = alg 𝓜
+
+    unfold-⟦_⟧ᴹ : ∀ t → ⟦ t ⟧ᴹ ≈ᴹ algᴹ (map η t)
+    unfold-⟦ var a ⟧ᴹ = begin η a ∎
+    unfold-⟦ s ∙ t ⟧ᴹ = congᴹ unfold-⟦ s ⟧ᴹ unfold-⟦ t ⟧ᴹ
+
+    cong-⟦_⟧ : ∀ {s t} → s ≈ t → ⟦ s ⟧ᴹ ≈ᴹ ⟦ t ⟧ᴹ
+    cong-⟦ var r ⟧ = cong-η r
+    cong-⟦ s ∙ t ⟧ = congᴹ cong-⟦ s ⟧ cong-⟦ t ⟧
+
+    isRelHomomorphismᴹ : IsRelHomomorphism _≈_ _≈ᴹ_ ⟦_⟧ᴹ
+    isRelHomomorphismᴹ = record { cong = cong-⟦_⟧ }
+
+    setoidHomomorphismᴹ : SetoidHomomorphism FA setoidᴹ
+    setoidHomomorphismᴹ = record { ⟦_⟧ = ⟦_⟧ᴹ ; isRelHomomorphism = isRelHomomorphismᴹ }
+
+    isMagmaHomomorphismᴹ : IsMagmaHomomorphism rawMagma rawMagmaᴹ ⟦_⟧ᴹ
+    isMagmaHomomorphismᴹ = record { isRelHomomorphism = isRelHomomorphismᴹ
+                                  ; homo = λ s t → begin ⟦ s ⟧ᴹ ∙ᴹ ⟦ t ⟧ᴹ ∎ }
```

Comment:
is the begin and end really needed?

Reply:
Interestingly it did seem to be, but I will look again. I was torn between and anonymous appeal to `refl` (and hence having to open a structure enough to expose that field, again with a sufficiently standardised-apart name) or else to make a trivial bit of 'equational reasoning' explicit so that the term is exhibited. The typechecker was grumbling about insufficiently specified field projection names at this point, and this was one way out of the bottle. But it's a bit of a wart, so I'll rethink/reconsider.

Reply:
Now replaced by appeal to `refl` instead.